components/ai_chat/browser/ai_chat_tab_helper.cc

```diff
@@ -320,6 +330,189 @@ void AIChatTabHelper::OnAPISuggestedQuestionsResponse(
+  } else if (is_video_) {
+    system_message = base::ReplaceStringPlaceholders(
+        l10n_util::GetStringUTF8(IDS_AI_CHAT_LLAMA2_SYSTEM_MESSAGE_VIDEO),
+        {article_text_}, nullptr);
+  } else {
+    system_message = base::ReplaceStringPlaceholders(
+        l10n_util::GetStringUTF8(IDS_AI_CHAT_LLAMA2_SYSTEM_MESSAGE_ARTICLE),
+        {article_text_}, nullptr);
+  }
+
+  auto conversation_history = GetConversationHistory();
+  // If there's no conversation history, then we just send a (partial)
+  // first sequence.
+  if (conversation_history.size() == 0) {
+    return BuildLlama2FirstSequence(system_message, user_message,
+                                    absl::nullopt);
+  }
+
+  // Use the first two messages to build the first sequence,
+  // which includes the system prompt.
+  std::string prompt =
+      BuildLlama2FirstSequence(system_message, conversation_history[0].text,
+                               conversation_history[1].text);
+
+  // Loop through the rest of the history two at a time building subsequent
+  // sequences.
+  for (size_t i = 2; i + 1 < conversation_history.size(); i += 2) {
+    const std::string& prev_user_message = conversation_history[i].text;
+    const std::string& assistant_message = conversation_history[i + 1].text;
+    prompt +=
+        BuildLlama2SubsequentSequence(prev_user_message, assistant_message);
+  }
+
+  // Build the final subsequent exchange using the current turn.
+  prompt += BuildLlama2SubsequentSequence(user_message, absl::nullopt);
+
+  // Trimming recommended by Meta
+  // https://huggingface.co/meta-llama/Llama-2-13b-chat#intended-use
+  prompt = base::TrimWhitespaceASCII(prompt, base::TRIM_ALL);
+  return prompt;
+}
+
+std::string AIChatTabHelper::BuildLlama2InstructionPrompt(
+    const std::string& instruction) {
+  return base::ReplaceStringPlaceholders(
+      R"( $1 $2 $3 )",
+      {ai_chat::kLlama2BIns, instruction, ai_chat::kLlama2EIns}, nullptr);
+}
+
+std::string AIChatTabHelper::BuildLlama2GenerateQuestionsPrompt(
+    bool is_video,
+    const std::string content) {
+  std::string content_template;
+  if (is_video) {
+    content_template =
+        l10n_util::GetStringUTF8(IDS_AI_CHAT_LLAMA2_GENERATE_QUESTIONS_VIDEO);
+  } else {
+    content_template =
+        l10n_util::GetStringUTF8(IDS_AI_CHAT_LLAMA2_GENERATE_QUESTIONS_ARTICLE);
+  }
+
+  const std::string& user_message =
+      base::ReplaceStringPlaceholders(content_template, {content}, nullptr);
+
+  return BuildLlama2FirstSequence(
+      l10n_util::GetStringUTF8(
+          IDS_AI_CHAT_LLAMA2_SYSTEM_MESSAGE_GENERATE_QUESTIONS),
+      user_message, absl::nullopt);
+}
+
+std::string AIChatTabHelper::BuildLlama2FirstSequence(
+    const std::string& system_message,
+    const std::string& user_message,
+    absl::optional<std::string> assistant_response) {
+  // Generates a partial sequence if there is no assistant_response:
+
+  // <s> [INST] <<SYS>>
```

Comment:
What did you use as reference for the format? I thought we wouldn't have needed the `[INST]` tags around the system prompt, but only `<<SYS>>` tags

Reply:
and what about the <s> tags?

Reply:
I used https://github.com/facebookresearch/llama/blob/6c7fe276574e78057f917549435a2554000a876d/llama/generation.py#L213 and https://huggingface.co/spaces/huggingface-projects/llama-2-13b-chat/blob/98df5b49e7a4fa1b3e406f5cdcb0fcb9fc8d6c75/model.py#L24 as reference among others. The template is not well documented and I've seen a number of variations.  I think this is close to correct.

Reply:
> Do we need the [INST] tags around the system prompt? Or only <<SYS>> tags

My understanding is yes.

> and what about the `<s>` tags?

I believe those wrap and exchange i.e. they encapsulate a user message and the assistants response.  So there there will be many `<s>..</s>` tags in a prompt which includes a long conversation history.  The  `</s>` tag is omitted for the last dialogue because we don't have the assistant's response yet

Reply:
Couldn't this be handled on the Anthropic-compatible API side?

Reply:
The Anthropic API expects an already formatted prompt and usually prompts need formatting to be done over different components [context, history, query, task_type], so it might be better handled by the client until we switch to server-side prompt building (which will require us to send to the server the conversation components rather than the already formatted prompt)